BIM: Hide all objects that aren't selected during Isolate

tetektoza opened the pull request:
Currently if user selects an item and does `Isolate` operation on it in BIM, everything is being hidden if the item was inside another container (like Building).

This is because we are prioritizing on hiding parents of current object if they are not the selected ones, which is causing the child of the parent to be hidden as well (duh).

So, this patch fixes isolate method to hide all other parents and their childs. If they are not a parent of our child under selection -> hide them and proceed further. If we are processing parent of our selection, just hide anything on current level that is not our selection while leaving parent intact.

Demo:

https://github.com/user-attachments/assets/2154df5b-b386-4b2d-86fe-e65c599abfb3


FYI: @semhustej @furgo16 

Resolves: https://github.com/FreeCAD/FreeCAD/issues/20886

furgo16 commented:
I'll test it this evening, thank you!

semhustej commented:
Looks good to me! I'll test when it is in the weekly builds.

furgo16 commented:
I've just built the branch and tested it. I used the `BIM Example.FCStd` example file that can be loaded from the Start page's `Examples` section. What I observed:

- Isolating by right-clicking on a level in the BIM Views panel works once, then it stops working. Stops working meaning: select another level in the BIM Views panel, right-click on Isolate, and then nothing happens.
- It takes quite a long time to complete the operation, whereas in comparison double-clicking on a level in the `BIM Views` panel is almost instantaneous. My guess is that double-clicking hides/shows the parent only, whereas right-clicking and executing the command hides all children one by one.

tetektoza commented:
@furgo16 
For the first point - I guess the solution would be to make all parents visible first before doing the operation, this will guarantee that we will get into a "reset" state.

For the second point, I will see what I can do with the performance tomorrow.

yorikvanhavre commented:
This seems to make sense. Waiting for @tetektoza 's feedback.

tetektoza commented:
@furgo16 So here's an actual difference between my implementation of Isolate and double-click on a level:

https://github.com/user-attachments/assets/3d585b14-914d-4d76-905c-1f920a0a965b

If you double click, you just hide the parent of the selected entity, but if you use my implementation of Isolate - you hide all other parents and their children. That's why it might be slower.

semhustej commented:
> SO, my question is - do we expect `Isolate` to work as double click? - only hide the selected entity's parent (and nothing else). OR do we want it to work as I've described above (and done so far in this PR)?

I have no clear opinion on this one.

- having two different isolate behaviors might be confusing to new users
- having two different isolate behaviors might be beneficial to experienced users
- having two different isolate behaviors is not good from the software maintanance perspective

semhustej commented:
> @semhustej it looks like double left click is something completely different than Isolate command, otherwise you would see it behaving the same way.

I meant 'two isolate behaviors' as of 'isolate' and 'double click'.

tetektoza commented:
Let me know what you want to do here. I can change one option to work like another. Even before this commit there was already a gap in usage between double left click and Isolate.

semhustej commented:
> Let me know what you want to do here. I can change one option to work like another. Even before this commit there was already a gap in usage between double left click and Isolate - because I guess double left click wasn't meant to be a Isolate alternative, just a wuick way to hide other parents on the same tree level.

Well, than I would let the PR as it is. It seems there was conscious reasoning behind having 'Isolate' behave in a different way than 'double click'.

But let's wait a bit and see whether others have something to say about it.

furgo16 commented:
My concern with the current Isolate implementation, while conceptually correct, is the performance. This is not a critique to this PR, though.

If I speak for myself, faced with the Isolate operation to take quite a few seconds vs the double-click operation to do (only apparently to the eyes of the user) the same thing, I'd never use isolate. I'd use double-click instead. There is also the perception aspect that will associate the operation's slowness with FreeCAD being slow in general.

tetektoza commented:
@furgo16 Are you able to retest this one and give your insights? It works kinda differently now. Since we get the items from the tree in `update` anyways, we will just cache those in member variable and reuse them in `isolate`. This way we don't have to do any  recursive tree traversal, which I think was the most time consuming operation here.

Also, I've added your request:
>Isolating by right-clicking on a level in the BIM Views panel works once, then it stops working. Stops working meaning: select another level in the BIM Views panel, right-click on Isolate, and then nothing happens.

Before we run the real logic of `isolate`, we show all items beforehand so it gets into a "reset" state, you can check it.

One more thing is - this above done logic kinda really isolates the selected items, it means that it also hides the parent of the selected object but leaves the child visible. I think this approach is better, since `BIM Views` menu is giving an option to select a parent separately, so if anyone wants to have parent visible too, they can just select it additionally.

furgo16 commented:
I'll test this later on today and report back, thanks!

yorikvanhavre commented:
Waiting for @furgo16 's feedback

tetektoza commented:
@furgo16 My bad, sorry for your time. I didn't take into consideration that we have a second part of the tree which contains directories, etc. And it didn't have the same members as normal Level has for example, so I just changed the range in which we cache items from the tree. Should be fine now as I've also tested it. I've also added sanity null check to just prevent that in the future.

I've tested it and it's still a little bit slower than double-click, but you have to note that still, double-click LOOKS the same in this scenario, but it doesn't do the same normally, in different scenarios. Although I think it is quite fast now. You can check yourself.

furgo16 commented:
> @furgo16 My bad, sorry for your time. I didn't take into consideration that we have a second part of the tree which contains directories, etc. And it didn't have the same members as normal Level has for example, so I just changed the range in which we cache items from the tree. Should be fine now as I've also tested it. I've also added sanity null check to just prevent that in the future.

No worries, thanks for addressing this

> I've tested it and it's still a little bit slower than double-click, but you have to note that still, double-click LOOKS the same in this scenario, but it doesn't do the same normally, in different scenarios. Although I think it is quite fast now. You can check yourself.

I've tested it with the branch's HEAD, including the removal of the recompute (which I'd tend to agree is not necessary, as double-click does not recompute it either). Observations:

1. Isolate speed is now noticeably faster. Double-click is still snappier than Isolate. From my point of view though, performance-wise Isolate is now acceptable.
2. It would be great if others could also test it and weigh in. With the changes contained in one file, it's relatively easy to test by downloading and copying over the BimViews.py file to a local installation (for those who don't want/can't build the branch).
3. I think it's now more of a conceptual question: should Isolate hide all non-selected parent levels and their children, or should, as a shortcut, just hide the parents? @yorikvanhavre what are your thoughts?
4. Using the shipped `BIMExample.FCStd` I've noticed this behavior, which is the same for double-click and Isolate: only parent levels and children items are affected by the operation. If something is outside of a level, it won't be hidden or unhidden. This can be seen on `BIMExample.FCStd`: whenever double-clicking on a level on the BIM Views window, or doing Isolate, the terrain is always visible. I'd say it's up for discussion whether it (and in general objects outside levels) should be affected by Isolate. Conceptually, I'd think isolate means hide _everything_ except this level, but I don't have experience with what other BIM editors/IFC viewers do. It might be worth checking that, but it might be outside of the scope of this PR.
5. Also outside of the scope of this PR, but I thought I'd mention it so that it is recorded: unlike the tree view, you have to always first click on an item on the BIM Views panel to act on it. This leads to confusing behaviour whereby hovering an item would appear to select it (it is highlighted). However, if you then if you right-click on it, the context menu pops up and half a second later the focus jumps to the previously clicked item. I'll try to file an issue about it.

```
OS: Ubuntu 22.04.5 LTS (ubuntu:GNOME/ubuntu/wayland)
Architecture: x86_64
Version: 1.1.0dev.41425 +4 (Git)
Build date: 2025/05/05 23:55:28
Build type: Unknown
Branch: fix/fix_isolate_hiding_all_elements
Hash: f6d10d4b8f02b022ba2bba2341ddd62b2a5b5b7d
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, boost 1_74, Eigen3 3.4.0, PySide 5.15.2
shiboken 5.15.2, SMESH 7.7.1.0, xerces-c 3.2.3, IfcOpenShell 0.8.1.post1, OCC 7.8.1
Stylesheet/Theme/QtStyle: FreeCAD Light.qss/FreeCAD Light/Fusion
Logical DPI/Physical DPI/Pixel Ratio: 96/159.34/1
Installed mods: 
  * OpenTheme 2024.9.1
  * fasteners 0.5.38
```

tetektoza commented:
I just wanted to add on 5th point that you've stated - this is because of the timer that 'update' is working on, which is messing with selection most of the times. It disappears, etc. etc. I can't find this issue now, but this is pretty related to latest issue reported by Tom, where he reported it's not possible to press CTRL and select multiple items. I'm working on removing this timer but it requires a little bit of work, so both of those problems may be gone soon after my PR will be posted.

furgo16 commented:
Thank you. I found the issue: https://github.com/FreeCAD/FreeCAD/issues/21024

furgo16 commented:
The timer might also be responsible for https://github.com/FreeCAD/FreeCAD/issues/17415 now that I'm re-reading it.

semhustej commented:
I tried testing by downloading the `BimViews.py` file. These are my findings:

A. as for speed I don't have any large dataset to test with. Speed using BIMexample and few other small datasets I tried seem to be not a problem with this version of the PR. Thanks for the optimization of this tool, as performance is an issue when using FreeCAD BIM.

B. I tested also with native IFC and hybrid projects, the functionality works OK

C. I have noticed, that isolating `Building` object hides levels, which are children of this `Building` object. Logic would tell me, that isolating the building would mean having visible the `Building` object + all the children (Levels) that this building contains (user thinking: I want to isolate this whole building). What are your opinions on this?
I think this is not a crucial point, so if it would take lot of time to code, it is not worth the effort at this time.

> 3\. I think it's now more of a conceptual question: should Isolate hide all non-selected parent levels and their children, or should, as a shortcut, just hide the parents? @yorikvanhavre what are your thoughts?

I would vote for hiding non selected parent levels. If you have two buildings in your project and select to isolate one level in one building, there is no reason to have the other building visible. The current PR behavior seems fine to me. Hopefully in the future, user will be able to select more that one object to isolate from BIM_Views.
Additionally I think it is more of a academic question at this point as I currently don't see many users having multiple buildings in one project.



> 4\. I've noticed this behavior, which is the same for double-click and Isolate: only objects inside the BIM Views tree are affected by the operation. If something is outside the BIM Views panel's tree, it won't be hidden or unhidden.

I noticed this too. Conceptually I would like to see everything but the selected Level hidden. But practically this would make quite a big mess out of the rest of the project, as without some kind of `Un-Isolate` tool, user would only have option to turn back everything on, or manually select objects to be turned on. This would also close tabs of pages, which are currently open. So in my opinion at this stage the current behavior of the PR is correct.



> 5\. Also outside of the scope of this PR, but I thought I'd mention it so that it is recorded: unlike the tree view, you have to always first click on an item on the BIM Views panel to act on it. This leads to confusing behavior whereby hovering an item would appear to select it (it is highlighted)

This is a big usability issue and I can see lot of users having problems with this behavior. @tetektoza  as this is not a topic of this PR, should I report this as an independent issue or would this be solved together with https://github.com/FreeCAD/FreeCAD/issues/21024?

**Summary**
Except for question C, this PR seems to be ready in my opinion.

tetektoza commented:
@semhustej About point C - I've uploaded a patch. I've done small iteration parent traversal, which MAY add a LITTLE bit of overhead additionally if user has selected "Building" (this doesn't happen for other BIM entities as an optimization). The tree traversal has been done because I didn't know if someone will ever want to have a structure like:

Building1 -> Building2 -> Level1

Then Qt doesn't provide an easy way to check if Building1 is an ancestor/parent of Level1, thus I've added this traversal. It shouldn't be really problematic. You can test this out guys.

https://github.com/user-attachments/assets/b1491b57-c234-4490-b3b5-2cce710f3a4f

@semhustej You can check if that's what you meant in the above demo, but I also encourage for testing :)

semhustej commented:
> 
> @semhustej You can check if that's what you meant in the above demo, but I also encourage for testing :)

Thanks for providing the demonstration. I did a test with downloaded library. It works well. 

I encountered inconsistent behavior of isolating the building container based on the state of levels inside of the building at the beginning of the command. The inconsistency occurs with objects inside of the level (eg. walls, slabs, roofs, etc.)

I come form the basic understanding, that when some objects in level are hidden and the `isolate` command is used on the level, all objects in the level are made visible.

There are four cases that I tried:
**Case 1** Isolating a level

1a. At the beginning, the level is visible and an object (roof) inside of the level is visible
Using `isolate` on the level results in the roof being visible. This is OK.

1b. At the beginning, the level is visible and an object (roof) inside of the level is hidden
Using `isolate` on the level results in the roof being visible. This is OK.

**Case 2** Isolating a building with levels inside of it

2a. At the beginning, the building is visible, level is visible too and an object (roof) inside of the level is visible
Using `isolate` on the building results in the roof being visible. This is OK.

2b. At the beginning, the building is visible level is not visible and an object (roof) inside of the level is hidden
Using `isolate` on the building results in the roof being visible. This is OK.

2c. At the beginning, the building is visible **level is visible** and an object (roof) inside of the level is hidden
Using `isolate` on the building results in **the roof being hidden**. This is not OK as it is not consistent with other `isolate` results

I think the reason in 2c the roof remains hidden is that the algorithm sees visible level and does not do anything about it. 

See behavior in video (the problematic part is the roof), I hope it makes to problem clear together with the above description. The last one is the problematic behavior:

https://github.com/user-attachments/assets/8725d9e6-a159-4587-98fa-bdadeb2bf109

This is 2b - it works when the level is hidden at the beginning:


https://github.com/user-attachments/assets/b21e0024-072c-4de4-bd4b-465ce77aeec2

semhustej commented:
>  Probably a good hack here is to make those hidden and then visible once again (to trick them for refresh without using recompute)
> 

I thought of this but that would mean hide/unhide for each level. This would lead to reduced performance I guess. That's why I posted the 'brainstorming' part. 

Let's see if you figure something out.

tetektoza commented:
@semhustej So I've stayed with the primary idea, let me explain. The Visibility is not currently updated, because it seems that we update Visibility of direct child, so if you have something like this:

Building (visibility off) -> Level (visibility on) -> Roof (visibility off)

Then Roof will get updated visibility to `on`, but if you have something like this:

Building (visibility off) -> Level (visibility on) -> Group (on) -> Roof (visibility off)

Then it will not get updated. I guess that direct child gets an update and if nothing changed in terms of visibility it looks like it's not propagated further down the chain. It would get updated Visibility property, BUT we'd have to iterate through the "Model" tree (through each item and update it accordingly), instead of our Bim Views tree object. This means, that I'd have to add ADDITIONAL tree processing in either `update` or `isolate`, which is generally adding overhead, but on top of that, we'd hide/unhide a lot of items that we don't even want to unhide because they are unrelated to the BIM Views Tree. My guess is that it would be generally slower than unhide/show all relevant items. So I've done this little hack that we spoke about and tested on BIMExample, but I didn't see a lot of lost performance. You can check yourself.

semhustej commented:
Thanks for the explanation @tetektoza . I think it's a good solution, small performance drop when isolating the building is worth the consistent behavior.

I tested the latest library and it works well in my tests. In my opinion it is ready to be merged (from a user perspective).

yorikvanhavre commented:
Great! I'll have a last look and merge.

yorikvanhavre commented:
Sorry for my absence recently guys (baby's fault! :sweat_smile: ), and thanks to all of you for the hard work!

This seems to work well, so I'm going to merge it.

Regarding annexed questions about 1) performance and 2) difference with simply double-clicking a level, I would say the main issue is we have several concurrent systems in BIM, not only for isolating but in many other places too. It is not clear in my mind (and I guess not only me) what is the best path here. So probably, the best path to discover the best path, is to work on making things better wherever we can, and the best solution will naturally emerge.

semhustej commented:
I did few tests in the weekly build and it seems to work OK. Thank you!

```
OS: Windows 10 build 19045
Architecture: x86_64
Version: 1.1.0dev.41694 (Git)
Build date: 2025/05/13 12:46:54
Build type: Release
Branch: main
Hash: 5e624a4775c993a3d940f37876840c65ac0cc60c
Python 3.11.12, Qt 6.8.3, Coin 4.0.3, Vtk 9.3.1, boost 1_86, Eigen3 3.4.0, PySide 6.8.3
shiboken 6.8.3, xerces-c 3.2.5, IfcOpenShell 0.8.2, OCC 7.8.1
Locale: English/United States (en_US)
Stylesheet/Theme/QtStyle: FreeCAD Light.qss/FreeCAD Light/
Logical DPI/Physical DPI/Pixel Ratio: 96/126.557/1.25
Installed mods: 
  * Reporting
  * Telemetry 1.0.1

```
